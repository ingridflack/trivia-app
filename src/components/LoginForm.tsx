import { FormControl, FormLabel, Input } from "@chakra-ui/react";

export default function LoginForm() {
  return (
    <FormControl>
      <FormLabel>Email address</FormLabel>
      <Input type="email" />
      <FormLabel marginTop='16px'>Password</FormLabel>
      <Input type="password" />
    </FormControl>
  );
}
