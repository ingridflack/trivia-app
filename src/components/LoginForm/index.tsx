import {
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  Stack,
} from "@chakra-ui/react";
import { FieldErrors, UseFormRegister } from "react-hook-form";
import { LoginFormValues } from "../../pages/Login";
import { BaseSyntheticEvent } from "react";

interface LoginFormProps {
  onSubmit: (
    e?: BaseSyntheticEvent<object, unknown, unknown> | undefined
  ) => Promise<void>;
  errors: FieldErrors<LoginFormValues>;
  register: UseFormRegister<LoginFormValues>;
  isSubmitting: boolean;
}

export default function LoginForm({
  onSubmit,
  errors,
  register,
  isSubmitting,
}: LoginFormProps) {
  return (
    <Stack
      as="form"
      onSubmit={onSubmit}
      gap="16px"
      maxWidth="400px"
      role="form"
    >
      <FormControl isInvalid={!!errors.email}>
        <FormLabel htmlFor="email">Email address</FormLabel>
        <Input
          type="email"
          placeholder="Enter your email"
          id="email"
          {...register("email", {
            required: "Email is required.",
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: "Invalid email address",
            },
          })}
        />
        <FormErrorMessage>
          {errors.email && errors.email.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.password}>
        <FormLabel htmlFor="password">Password</FormLabel>
        <Input
          type="password"
          placeholder="Enter your password"
          id="password"
          {...register("password", {
            required: "Password is required.",
          })}
        />
        <FormErrorMessage>
          {errors.password && errors.password.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        colorScheme="purple"
        size="lg"
        width="100%"
        isLoading={isSubmitting}
        type="submit"
      >
        Sign in
      </Button>
    </Stack>
  );
}
