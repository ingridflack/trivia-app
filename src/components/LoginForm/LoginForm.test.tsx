import {
  render,
  screen,
  fireEvent,
  waitFor,
  cleanup,
} from "@testing-library/react";
import { afterEach, describe, expect, it, vi } from "vitest";

import LoginForm from ".";

const mockOnSubmit = vi.fn();

const setup = (errors = {}, isSubmitting = false) => {
  const register = vi.fn();
  render(
    <LoginForm
      onSubmit={mockOnSubmit}
      errors={errors}
      register={register}
      isSubmitting={isSubmitting}
    />
  );
};

describe("LoginForm", () => {
  setup();

  afterEach(() => {
    cleanup();
  });

  it("should match the snapshot", () => {
    const { container } = render(
      <LoginForm
        onSubmit={mockOnSubmit}
        errors={{}}
        register={vi.fn()}
        isSubmitting={false}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("displays email error message when email is invalid", () => {
    setup({ email: { message: "Invalid email address" } });
    expect(screen.getByText(/invalid email address/i)).toBeTruthy();
  });

  it("displays password error message when password is required", () => {
    setup({ password: { message: "Password is required." } });
    expect(screen.getByText(/password is required/i)).toBeTruthy();
  });

  it("calls onSubmit when form is submitted", async () => {
    setup();
    fireEvent.submit(screen.getByRole("form"));
    await waitFor(() => expect(mockOnSubmit).toHaveBeenCalled());
  });

  it("button is loading when isSubmitting is true", () => {
    setup({}, true);
    expect(screen.getByRole("button", { name: /sign in/i })).toHaveProperty(
      "disabled",
      true
    );
  });
});
