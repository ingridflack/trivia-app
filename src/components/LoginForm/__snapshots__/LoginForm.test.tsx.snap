// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > should match the snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-2umoaz"
    role="form"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="email"
        id="field-:r2:-label"
      >
        Email address
      </label>
      <input
        class="chakra-input css-0"
        id="email"
        placeholder="Enter your email"
        type="email"
      />
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="password"
        id="field-:r3:-label"
      >
        Password
      </label>
      <input
        class="chakra-input css-0"
        id="password"
        placeholder="Enter your password"
        type="password"
      />
    </div>
    <button
      class="chakra-button css-uxt1e8"
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
